File: widgetssdk/src/main/java/com/glia/widgets/filepreview/domain/usecase/DownloadFileUseCase.java
package com.glia.widgets.filepreview.domain.usecase;

import com.glia.androidsdk.chat.AttachmentFile;
import com.glia.widgets.filepreview.data.GliaFileRepository;
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException;
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException;

import io.reactivex.Completable;

public class DownloadFileUseCase {
    private final GliaFileRepository fileRepository;

    public DownloadFileUseCase(GliaFileRepository fileRepository) {
        this.fileRepository = fileRepository;
    }

    public Completable execute(AttachmentFile file) {
        if (file == null || file.getName().isEmpty())
            return Completable.error(new FileNameMissingException());
        if (file.isDeleted())
            return Completable.error(new RemoteFileIsDeletedException());

        return this.fileRepository
                .downloadFileFromNetwork(file);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/filepreview/domain/usecase/DownloadFileUseCase.kt
package com.glia.widgets.filepreview.domain.usecase

import com.glia.androidsdk.chat.AttachmentFile
import com.glia.widgets.core.secureconversations.domain.ManageSecureMessagingStatusUseCase
import com.glia.widgets.filepreview.data.GliaFileRepository
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException
import io.reactivex.rxjava3.core.Completable

internal class DownloadFileUseCase(
    private val fileRepository: GliaFileRepository,
    private val manageSecureMessagingStatusUseCase: ManageSecureMessagingStatusUseCase
) {
    operator fun invoke(file: AttachmentFile): Completable = when {
        file.name.isEmpty() -> Completable.error(FileNameMissingException())
        file.isDeleted -> Completable.error(RemoteFileIsDeletedException())
        else -> fileRepository.downloadFileFromNetwork(manageSecureMessagingStatusUseCase.shouldUseSecureMessagingEndpoints, file)
    }
}
